src/Features/Core/Portable/Diagnostics/IDiagnosticAnalyzerService.cs

```diff
@@ -22,8 +23,22 @@ internal interface IDiagnosticAnalyzerService : IWorkspaceService
     /// </remarks>
     void RequestDiagnosticRefresh();
 
-    /// <inheritdoc cref="IRemoteDiagnosticAnalyzerService.ForceAnalyzeProjectAsync"/>
-    Task<ImmutableArray<DiagnosticData>> ForceAnalyzeProjectAsync(Project project, CancellationToken cancellationToken);
+    /// <summary>
+    /// The default analyzer filter that will be used in functions like <see cref="GetDiagnosticsForIdsAsync"/> if
+    /// no filter is provided.  The default filter has the following rules:
+    /// <list type="number">
+    /// <item>The standard compiler analyzer will not be run if the compiler diagnostic scope is <see cref="CompilerDiagnosticsScope.None"/>.</item>
+    /// <item>A regular analyzer will not be run if <see cref="ProjectState.RunAnalyzers"/> is false.</item>
+    /// <item>A regular analyzer will not be run if if the background analysis scope is <see cref="BackgroundAnalysisScope.None"/>.</item>
+    /// <item>If a set of diagnostic ids are provided, the analyzer will not be run unless it declares at least one
+    /// descriptor in that set.</item>
+    /// <item>Otherwise, the analyzer will be run</item>
+    /// </list>
+    /// </summary>
+    /// <param name="additionalFilter">An additional filter that can accept or reject analyzers that the default
+    /// rules have accepted.</param>
+    Func<DiagnosticAnalyzer, bool> GetDefaultAnalyzerFilter(
+        Project project, ImmutableHashSet<string>? diagnosticIds, Func<DiagnosticAnalyzer, bool>? additionalFilter = null);
```

Comment:
It does feel a little weird that we pass another filter to retrieve the default filter.  Wondering if places that use the `additionalFilter` should instead be creating their own single filter (which may wrap a call to the default filter), instead of it being passed in?

It might make usages more clear that even though they are getting the default filter, they may not be using the default filter behavior.

Reply:
Hrmm... i'll see if there's a better way to do this.  Originally, i just returned the filtetr, and then the callers had to combine it with theirs.

In practice, the pattern was *always* teh same.  So having this pattern just helped encode both the idea of:

1. get the default rules
2. get the default rules, augmented with some other set.

So with this, and the ability for someone to pass a callback to GetDiagnostics meant we could handle all 3 cases:

1. get the default rules
2. get the default rules, augmented with some other set.
3. use an entirely custom set of rules

But yeah, i agree it's not necessarily stunning here...